pkg/services/sqlstore/migrations/accesscontrol/team_membership.go

```diff
@@ -0,0 +1,209 @@
+	case models.PERMISSION_ADMIN:
+		return []accesscontrol.Permission{
+			{Action: "teams:delete", Scope: teamIDScope},
+			{Action: "teams:read", Scope: teamIDScope},
+			{Action: "teams:write", Scope: teamIDScope},
+			{Action: "teams.permissions:read", Scope: teamIDScope},
+			{Action: "teams.permissions:write", Scope: teamIDScope},
+		}
+	default:
+		return []accesscontrol.Permission{}
+	}
+}
+
+func (p *teamPermissionMigrator) getUserRoleByOrgMapping(sess *xorm.Session) (map[int64]map[int64]string, error) {
+	var orgUsers []*models.OrgUserDTO
+	if err := sess.SQL(`SELECT * FROM org_user`).Cols("org_user.org_id", "org_user.user_id", "org_user.role").Find(&orgUsers); err != nil {
+		return nil, err
+	}
+
+	userRolesByOrg := map[int64]map[int64]string{}
+
+	// Loop through users and organise them by organization ID
+	for _, orgUser := range orgUsers {
+		orgRoles, initialized := userRolesByOrg[orgUser.OrgId]
+		if !initialized {
+			orgRoles = map[int64]string{}
+		}
+
+		orgRoles[orgUser.UserId] = orgUser.Role
+		userRolesByOrg[orgUser.OrgId] = orgRoles
+	}
+
+	return userRolesByOrg, nil
+}
+
+// migrateMemberships generate managed permissions for users based on their memberships to teams
+func (p *teamPermissionMigrator) migrateMemberships(sess *xorm.Session) error {
+	userRolesByOrg, err := p.getUserRoleByOrgMapping(sess)
+	if err != nil {
+		return err
+	}
+
+	var teamMemberships []models.TeamMember
+	if err := sess.SQL(`SELECT * FROM team_member`).Find(&teamMemberships); err != nil {
+		return err
+	}
+
+	userPermissionsByOrg := map[int64]map[int64][]accesscontrol.Permission{}
+
+	// Loop through memberships and generate associated permissions
+	for _, m := range teamMemberships {
+		// Downgrade team permissions if needed - only organisation admins or organisation editors (when editorsCanAdmin feature is enabled)
+		// can access team administration endpoints
+		if m.Permission == models.PERMISSION_ADMIN {
+			if userRolesByOrg[m.OrgId][m.UserId] == string(models.ROLE_VIEWER) || (userRolesByOrg[m.OrgId][m.UserId] == string(models.ROLE_EDITOR) && !p.editorsCanAdmin) {
+				m.Permission = 0
+				_, err := sess.Cols("permission").Where("org_id=? and team_id=? and user_id=?", m.OrgId, m.TeamId, m.UserId).Update(m)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		userPermissions, initialized := userPermissionsByOrg[m.OrgId]
+		if !initialized {
+			userPermissions = map[int64][]accesscontrol.Permission{}
+		}
+		userPermissions[m.UserId] = append(userPermissions[m.UserId], p.mapPermissionToFGAC(m.Permission, m.TeamId)...)
+		userPermissionsByOrg[m.OrgId] = userPermissions
+	}
+
+	// Loop through generated permissions and store them
+	for orgID, userPermissions := range userPermissionsByOrg {
+		for userID, permissions := range userPermissions {
+			roleID, err := p.getOrCreateManagedRole(sess, orgID, userID, fmt.Sprintf("managed:users:%d:permissions", userID))
```

Comment:
(Y org x Z user) insertions on the `user_role` and `role` tables. Is it possible to batch these?